funsor/einsum/numpy_map.py

```diff
@@ -0,0 +1,55 @@
+# Copyright (c) 2017-2019 Uber Technologies, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+import operator
+
+from functools import reduce
+
+from pyro.ops import packed
+from pyro.ops.einsum.adjoint import Backward, einsum_backward_sample, transpose, unflatten
+from pyro.ops.einsum.util import Tensordot
+
+
+class _EinsumBackward(Backward):
+    def __init__(self, operands, argmax):
+        self.operands = operands
+        self.argmax = argmax
+
+    def process(self, message):
+        sample1 = self.argmax
+        sample2 = message
+        return einsum_backward_sample(self.operands, sample1, sample2)
+
+
+def einsum(equation, *operands):
+    """
+    Forward-max-sum backward-argmax implementation of einsum.
+    This assumes all operands have a ``._pyro_dims`` attribute set.
+    """
+    equation = packed.rename_equation(equation, *operands)
+    inputs, output = equation.split('->')
+    any_requires_backward = any(hasattr(x, '_pyro_backward') for x in operands)
+
+    contract_dims = ''.join(sorted(set().union(*(x._pyro_dims for x in operands)) - set(output)))
+    dims = output + contract_dims
+    result = reduce(operator.add, packed.broadcast_all(*operands, dims=dims))
```

Comment:
It looks like line 29 (`packed.rename_equation`) and line 35 (`packed.broadcast_all`) are the only places in this code where `ops.packed` is used, at least outside of the adjoint methods.  I suggest copying those two functions into this file or `util.py` and deleteting all of the adjoint-related code in this PR, including `adjoint.py`, `packed.py`, all of the `Backward` implementations and any `_pyro_backward` properties.  Don't worry about missing adjoint functionality for now; just delete or xfail any tests as necessary.

Reply:
Thanks! I'll add comments for missing points. :)